### PR TITLE
[quantization] Add quantization support for LRN and SoftMax in middle-end

### PR DESCRIPTION
**Pull request by aturetsk**
Add quantization support for LRN and SoftMax in GLOW middle-end, making possible for any back-and to implement them, if needed.

### CODE REVIEW

File: lib/Quantization/Quantization.cpp [hunk 1]
```diff
@@ -296,6 +300,30 @@ static Node *quantizeNode(Function *F, Node *node,
...
     break;
   }
+  case Kinded::Kind::LocalResponseNormalizationNodeKind: {
+    auto *LRN = cast<LocalResponseNormalizationNode>(node);
+    assert(quantizedInputs.size() == 1 && "Invalid number of inputs");
+    assert(qParams.size() == 1 && "Invalid number of quantized outputs");
+
+    // Note: This should either be lowered into an IntLookupTable, or
+    // implemented via a backend-specific Node/Inst.
+    quantizedNode = F->createLocalResponseNormalization(
```
Comment (rdzhabarov) [inline lib/Quantization/Quantization.cpp]: how LRN is supposed to be implemented via IntLookupTable?

File: lib/Quantization/Quantization.cpp [hunk 2]
```diff
@@ -296,6 +300,30 @@ static Node *quantizeNode(Function *F, Node *node,
...
+  }
+  case Kinded::Kind::SoftMaxNodeKind: {
+    auto *SMN = cast<SoftMaxNode>(node);
+    // SoftMax node has 2 inputs, but only one should be quantized.
+    assert(quantizedInputs.size() == 1 && "Invalid number of inputs");
+    assert(qParams.size() == 1 && "Invalid number of quantized outputs");
+
+    // Note: This should either be lowered into an IntLookupTable, or
+    // implemented via a backend-specific Node/Inst.
+    quantizedNode = F->createSoftMax(SMN->getName(), quantizedInputs[0],
```
Comment (rdzhabarov) [inline lib/Quantization/Quantization.cpp]: Do you have a plan to add softMax support to the interpreter?

File: lib/Quantization/Quantization.cpp [hunk 3]
```diff
@@ -99,6 +99,10 @@ static bool canBeQuantized(const Node *node) {
     auto *gather = cast<GatherNode>(node);
     return gather->getData().getElementType() == ElemKind::FloatTy;
   }
+  case Kinded::Kind::SoftMaxNodeKind: {
```
Comment (rdzhabarov) [inline lib/Quantization/Quantization.cpp]: as long as we quantize node, we should ensure proper support at least by interpreter backend.

File: lib/Quantization/Quantization.cpp (hunk 2 above)
Comment (nadavrot) [reply lib/Quantization/Quantization.cpp]: I am not sure how we can quantize SoftMax because the intermediate sum can be huge, but if some backends need it then we can go for it.

Comment (rdzhabarov) [reply lib/Quantization/Quantization.cpp]: I think @jspark1105 had some ideas about this. Where did we end up with the quantized softmax?

Comment (jspark1105) [reply lib/Quantization/Quantization.cpp]: We haven't quite finished our work on quantizing softmax. gemmlowp has a quantized softmax implementation: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/kernels/internal/optimized/optimized_ops.h#L3907

**Comment by rdzhabarov** [comment]
I think this PR is OK to ship to unblock you if you are blocked on this (without implementation of quantized softmax). Please, make sure to update the comments properly though. 
Another thing, can you ensure code coverage for these new operators. You can create mock backend and report isOpSupported for softmax and/or LRN in order to cover this path. See some examples in QuantizationTest.

**Comment by aturetsk** [comment]
> Please, make sure to update the comments properly though.

Yes, comments are a copy-paste error. I'll remove them.

> Another thing, can you ensure code coverage for these new operators. You can create mock backend and report isOpSupported for softmax and/or LRN in order to cover this path. See some examples in QuantizationTest.

I'm not sure if it can be done with current interfaces. quantizeFunction (which I need to test), has ExecutionEngine argument and takes Backend object from it. However, I can't initialize EE object with a custom back-end, because EE constructor and setBackend function take a enum rather than a Backend object.
Or am I missing something?

> I think this PR is OK to ship to unblock you if you are blocked on this (without implementation of quantized softmax). 

Yes, currently we are working on quantized LRN and Softmax in our back-end, so I had to add this code.
I just though that since the code is in middle end it might be a good idea to commit it in master (and it'd help to avoid rebase conflicts in the future). If you find it unacceptable to have the changes without the test coverage (due to the issues I described above), I can cancel the PR and keep the changes in our local repo.

**Comment by rdzhabarov** [comment]
> I'm not sure if it can be done with current interfaces. quantizeFunction (which I need to test), has ExecutionEngine argument and takes Backend object from it. However, I can't initialize EE object with a custom back-end, because EE constructor and setBackend function take a enum rather than a Backend object.
Or am I missing something?

Quantization conversion procedure relies on the isOpSupported from the ExecutionEngine and you can just pass MockEE to ```F2 = quantization::quantizeFunction(backendSpecificEE, QI, F2);```
Make mock to return true for certain operations and you should get F2 conversion as expected (Note, you'd need to make isOpSupported on the EE level virtual).

> I just though that since the code is in middle end it might be a good idea to commit it in master (and it'd help to avoid rebase conflicts in the future)

Certainly, let's do it. Do you think you could add a simple test for the conversion?

**Comment by aturetsk** [comment]
> Quantization conversion procedure relies on the isOpSupported from the ExecutionEngine and you can just pass MockEE to F2 = quantization::quantizeFunction(backendSpecificEE, QI, F2);
Make mock to return true for certain operations and you should get F2 conversion as expected (Note, you'd need to make isOpSupported on the EE level virtual).

But I can't derive MockEE from EE, since EE is 'final' class.
I was able to overcome the restriction by overloading setBackend method and allowing it to have a pointer to Backend object as argument (making it possible to pass a custom mock backend to it).
Are you OK with this approach?

**Comment by rdzhabarov** [comment]
I'm fine either with that (as long as backendKind is not kept in the internal state which is not) or just remove "final" as I do not see a strong reason to make that final as well.

**Comment by aturetsk** [comment]
@rdzhabarov, then let's stick to the current implementation.
Is there anything else I need to address before the patch can be committed?

**Comment by rdzhabarov** [comment]
oh, I did not realize that you updated the PR :) Will look @aturetsk

File: include/glow/ExecutionEngine/ExecutionEngine.h [hunk 4]
```diff
@@ -54,6 +54,9 @@ class ExecutionEngine final {
   // using this backend.
   void setBackend(BackendKind backendKind);
 
+  // Set the code generator to a custom \p backend.
```
Comment (rdzhabarov) [inline include/glow/ExecutionEngine/ExecutionEngine.h]: /// doxygen comments. If you could please fix the one above as well.

File: lib/ExecutionEngine/ExecutionEngine.cpp [hunk 5]
```diff
@@ -47,6 +47,12 @@ void ExecutionEngine::setBackend(BackendKind backendKind) {
   function_.reset();
 }
 
+// Set the code generator kind to \p backend.
```
Comment (rdzhabarov) [inline lib/ExecutionEngine/ExecutionEngine.cpp]: same here.

File: tests/unittests/quantizationTest.cpp [hunk 6]
```diff
@@ -577,6 +578,53 @@ TEST(Quantization, chooseQuantizationSymmetricWithUInt8) {
   EXPECT_NEAR(symmetricParams.scale, 16.0 / 255, 0.001);
 }
 
+/// This is a mock backend which extended support of quantizad operators.
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: typo in quantized.

File: tests/unittests/quantizationTest.cpp [hunk 7]
```diff
@@ -577,6 +578,53 @@ TEST(Quantization, chooseQuantizationSymmetricWithUInt8) {
...
+  F->createSave("ret", SM);
+
+  std::vector<NodeQuantizationInfo> QI{
+      {NodeQuantizationInfo::generateNodeOutputName(input->getName()),
+       {0.2f, 0}},
+      {NodeQuantizationInfo::generateNodeOutputName(LRN->getName()), {0.3f, 0}},
+      {NodeQuantizationInfo::generateNodeOutputName(SM->getName()), {0.4f, 0}},
+  };
+
+  F = quantization::quantizeFunction(EE, QI, F);
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: LGTM. Can you check that inputs to the LRN and SM is actually int8.

**Comment by aturetsk** [comment]
@rdzhabarov, thanks for review!
I fixed your remarks.

File: tests/unittests/quantizationTest.cpp [hunk 8]
```diff
@@ -577,6 +578,60 @@ TEST(Quantization, chooseQuantizationSymmetricWithUInt8) {
...
+
+  auto &mod = EE.getModule();
+  Function *F = mod.createFunction("main");
+
+  auto *input = mod.createVariable(ElemKind::FloatTy, {1, 10}, "input");
+  auto *selected = mod.createVariable(ElemKind::IndexTy, {1, 10}, "selected");
+  auto *LRN =
+      F->createLocalResponseNormalization("LRN", input, 2, 1.0, 0.0001, 0.75);
+  auto *SM = F->createSoftMax("softmax", LRN, selected);
+  F->createSave("ret", SM);
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: this is fine, but FYI, if you do not compile "save node" is not necessary.

File: tests/unittests/quantizationTest.cpp [hunk 9]
```diff
@@ -577,6 +578,60 @@ TEST(Quantization, chooseQuantizationSymmetricWithUInt8) {
...
+  std::vector<NodeQuantizationInfo> QI{
+      {NodeQuantizationInfo::generateNodeOutputName(input->getName()),
+       {0.2f, 0}},
+      {NodeQuantizationInfo::generateNodeOutputName(LRN->getName()), {0.3f, 0}},
+      {NodeQuantizationInfo::generateNodeOutputName(SM->getName()), {0.4f, 0}},
+  };
+
+  F = quantization::quantizeFunction(EE, QI, F);
+
+  auto *qLRN = cast<LocalResponseNormalizationNode>(F->getNodeByName("LRN1"));
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: "LRN1" and "softmax1" is very fragile (depends on the naming schema used, we could change that potentially etc).

Could you just iterate through the nodes in quantized F and check for LRN or SM nodes and then verify that they have quantized type input.

Comment (rdzhabarov) [reply tests/unittests/quantizationTest.cpp]: Otherwise, if you iterate there will be one non-quantized and one quantized node, also kinda messy...

**Comment by rdzhabarov** [comment]
Thanks @aturetsk!
